teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md

```diff
@@ -0,0 +1,544 @@
+
+The following are the defined schemas:
+```
+  - name: CHAINS-PYTHON_SDIST
+    type: object
+    description: |
+      * name: name or url of the source distribution
+      * digest: digest of the distribution
+      * alg: digest algorithm: sha256 etc.
+      * prov_type: 'input' or 'output', indicating provenance type.
+    properties:
+    name:
+      type: string
+    digest:
+      type: string
+    alg:
+      type: string
+    prov_type:
+      type: string
+```
+
+```
+  - name: CHAINS-PYTHON_BDIST
+    type: object
+    description: |
+      * name: name or url of the built distribution
+      * digest: digest of the distribution 
+      * alg: digest algorithm: sha256 etc.
+      * prov_type: 'input' or 'output', indicating provenance type.
+    properties:
+    name:
+      type: string
+    digest:
+      type: string
+    alg:
+      type: string
+    prov_type:
+      type: string
+```
+#### Maven
+
+Maven supports several for compiling and packaging. Compiling is done via `mvn compile` and produces `.class` files and packackging is done via `mvn package`, and [supports](https://maven.apache.org/pom.html#packaging) `pom`, `jar`, `maven-plugin`, `ejb`, `war`, `ear` and `rar`. Some plugins such as [Jib](https://github.com/GoogleContainerTools/jib/blob/master/jib-maven-plugin/README.md) can also help containerizing the compiled binaries into OCI images, but that can be covered by OCI image schema.
+
+Therefore the schema for Maven packages is:
+```
+  - name: CHAINS-MAVEN_PACKAGE
+    type: object
+    description: |
+      * name: name or url of the Maven package. Example: foo-v1.0.jar
+      * digest: digest of the package
+      * alg: digest algorithm: sha256 etc.
+      * prov_type: 'input' or 'output', indicating provenance type.
+    properties:
+    name:
+      type: string
+    digest:
+      type: string
+    alg:
+      type: string
+    prov_type:
+      type: string
+```
+
+#### Go
+
+TODO
+
+
+
+#### NodeJS
+
+TODO
+
+#### Parsing Results and Params from TaskRuns
+
+The defined results and params are in object types, and will be retrieved from TaskRun params and results. From the syntax of how the results / params look, the type of those, whether they are strings, objects or arrays can be determined. We need to validate the targets’ structures against defined result structures. Taking VCS as an example, when a `CHAINS-VCS` is found, the target needs to be an object, and it has to have at least the fields in defined `CHAINS-VCS` schema: `src_type`, `uri`, `alg` and `digest`. If additional fields are in the target, they will be ignored. The field validation rule follows the same pattern of duck typing mentioned in [TEP-0075](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md). These individual fields will also be validated against supported formats.
```

Comment:
Based on the discussion above, is the plan to still rely on user params or rather have the task output a result that expands the user input into a structured result that Chains can use?

Reply:
Thanks for reminding this! Now I removed all the proposals related to params.